File: sdk/python/workswithagents/deployment.py
"""
Agent Deployment Manifest — Cross-Layer
Docker Compose for agent fleets. Deploy from YAML.

Usage:
    from workswithagents import DeploymentManifest
    
    dm = DeploymentManifest.from_file("fleet.yaml")
    dm.validate()
    dm.deploy()
    dm.status()
"""

import json
import urllib.request
import urllib.error
from typing import Optional
from pathlib import Path

DEFAULT_API = "https://workswithagents.dev"


class DeploymentManifest:
    """Parse, validate, and deploy agent fleet manifests."""
    
    def __init__(self, manifest: dict, api_url: str = DEFAULT_API):
        self.manifest = manifest
        self.api = api_url.rstrip("/")
        self.fleet_id = None
    
    @classmethod
    def from_file(cls, path: str, api_url: str = DEFAULT_API) -> "DeploymentManifest":
        """Load manifest from YAML or JSON file."""
        content = Path(path).read_text()
        if path.endswith('.json'):
            import json
            return cls(json.loads(content), api_url)
        else:
            # Try YAML, fall back to JSON
            try:
                import yaml
                return cls(yaml.safe_load(content), api_url)
            except ImportError:
                # Minimal YAML parser for common cases
                return cls(_parse_simple_yaml(content), api_url)
    
    @classmethod
    def minimal(cls, name: str, agent_id: str, 
                capabilities: list, api_url: str = DEFAULT_API) -> "DeploymentManifest":
        """Create a minimal manifest for a single agent."""
        return cls({
            "manifest_version": "1.0.0-draft",
            "fleet": {
                "name": name,
                "agents": [{
                    "id": agent_id,
                    "type": "hermes",
                    "capabilities": capabilities
                }]
            }
        }, api_url)
    
    def validate(self) -> dict:
        """Validate manifest structure. Returns {valid, errors}."""
        errors = []
        mf = self.manifest
        
        if "manifest_version" not in mf:
            errors.append("Missing manifest_version")
        if "fleet" not in mf:
            errors.append("Missing fleet section")
            return {"valid": False, "errors": errors}
        
        fleet = mf["fleet"]
        if "name" not in fleet:
            errors.append("Missing fleet.name")
        if "agents" not in fleet or not fleet["agents"]:
            errors.append("Missing fleet.agents (at least one required)")
        
        for i, agent in enumerate(fleet.get("agents", [])):
            if "id" not in agent:
                errors.append(f"Agent[{i}]: missing id")
            if "capabilities" not in agent:
                errors.append(f"Agent[{i}]: missing capabilities")
        
        return {"valid": len(errors) == 0, "errors": errors}
    
    def deploy(self) -> dict:
        """Deploy fleet from manifest."""
        validation = self.validate()
        if not validation["valid"]:
            return {"status": "error", "errors": validation["errors"]}
        
        result = self._request("POST", "/v1/fleets/deploy", self.manifest)
        if "fleet_id" in result:
            self.fleet_id = result["fleet_id"]
        return result
    
    def status(self) -> dict:
        """Get fleet status."""
        if not self.fleet_id:
            return {"error": "Fleet not deployed. Call deploy() first."}
        return self._request("GET", f"/v1/fleets/{self.fleet_id}/status")
    
    def scale(self, agent_type: str, count: int) -> dict:
        """Scale an agent type up or down."""
        if not self.fleet_id:
            return {"error": "Fleet not deployed."}
        return self._request("POST", f"/v1/fleets/{self.fleet_id}/scale", {
            "agent_type": agent_type,
            "count": count
        })
    
    def _request(self, method: str, path: str, data: Optional[dict] = None) -> dict:
        url = f"{self.api}{path}"
        body = json.dumps(data).encode() if data else None
        req = urllib.request.Request(url, data=body, method=method)
        req.add_header("Content-Type", "application/json")
        try:
            with urllib.request.urlopen(req) as resp:
                return json.loads(resp.read())
        except urllib.error.HTTPError as e:
            return {"error": e.code, "message": e.read().decode()}


def _parse_simple_yaml(text: str) -> dict:
    """Minimal YAML parser for common manifest patterns. Falls back to JSON."""
    # Try JSON first
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        pass
    
    # Basic YAML: only handles simple key: value and nested dicts
    # For full YAML support: pip install pyyaml
    result = {}
    current = result
    stack = []
    indent_level = 0
    
    for line in text.split('\n'):
        stripped = line.strip()
        if not stripped or stripped.startswith('#'):
            continue
        
        # Calculate indent
        indent = len(line) - len(line.lstrip())
        
        if ':' in stripped:
            key, _, value = stripped.partition(':')
            key = key.strip()
            value = value.strip()
            
            if value == '':
                # Start of nested dict
                current[key] = {}
                stack.append(current)
                current = current[key]
            else:
                # Simple key: value
                # Try parsing value
                if value.lower() == 'true':
                    value = True
                elif value.lower() == 'false':
                    value = False
                elif value.isdigit():
                    value = int(value)
                elif value.startswith('"') and value.endswith('"'):
                    value = value[1:-1]
                current[key] = value
    
    return result

File: sdk/typescript/src/deployment.ts
/**
 * Agent Deployment Manifest — Cross-Layer (TypeScript)
 * Deploy agent fleets from YAML/JSON manifests.
 * Reference implementation. CC BY 4.0.
 */

const DEFAULT_API = "https://workswithagents.dev";

export interface DeploymentAgent {
  id: string;
  type?: string;
  capabilities: string[];
  [key: string]: unknown;
}

export interface DeploymentFleet {
  name: string;
  agents: DeploymentAgent[];
  [key: string]: unknown;
}

export interface DeploymentManifestData {
  manifest_version: string;
  fleet: DeploymentFleet;
  [key: string]: unknown;
}

export class DeploymentManifest {
  manifest: DeploymentManifestData;
  api: string;
  fleet_id: string | null = null;

  constructor(manifest: DeploymentManifestData, api_url: string = DEFAULT_API) {
    this.manifest = manifest;
    this.api = api_url.replace(/\/$/, "");
  }

  static fromObject(obj: Record<string, unknown>, api_url: string = DEFAULT_API): DeploymentManifest {
    return new DeploymentManifest(obj as DeploymentManifestData, api_url);
  }

  static minimal(name: string, agent_id: string, capabilities: string[], api_url: string = DEFAULT_API): DeploymentManifest {
    return new DeploymentManifest({
      manifest_version: "1.0.0-draft",
      fleet: { name, agents: [{ id: agent_id, type: "hermes", capabilities }] }
    }, api_url);
  }

  validate(): { valid: boolean; errors: string[] } {
    const errors: string[] = [];
    const mf = this.manifest;
    if (!mf.manifest_version) errors.push("Missing manifest_version");
    if (!mf.fleet) { errors.push("Missing fleet section"); return { valid: false, errors }; }
    if (!mf.fleet.name) errors.push("Missing fleet.name");
    if (!mf.fleet.agents?.length) errors.push("Missing fleet.agents (at least one required)");
    mf.fleet.agents?.forEach((agent, i) => {
      if (!agent.id) errors.push(`Agent[${i}]: missing id`);
      if (!agent.capabilities) errors.push(`Agent[${i}]: missing capabilities`);
    });
    return { valid: errors.length === 0, errors };
  }

  async deploy(): Promise<Record<string, unknown>> {
    const validation = this.validate();
    if (!validation.valid) return { status: "error", errors: validation.errors };
    const result = await this.request("POST", "/v1/fleets/deploy", this.manifest as unknown as Record<string, unknown>);
    if (result.fleet_id) this.fleet_id = result.fleet_id as string;
    return result;
  }

  async status(): Promise<Record<string, unknown>> {
    if (!this.fleet_id) return { error: "Fleet not deployed. Call deploy() first." };
    return this.request("GET", `/v1/fleets/${this.fleet_id}/status`);
  }

  async scale(agent_type: string, count: number): Promise<Record<string, unknown>> {
    if (!this.fleet_id) return { error: "Fleet not deployed." };
    return this.request("POST", `/v1/fleets/${this.fleet_id}/scale`, { agent_type, count });
  }

  private async request(method: string, path: string, data?: Record<string, unknown>): Promise<any> {
    const url = `${this.api}${path}`;
    const opts: RequestInit = { method, headers: { "Content-Type": "application/json" } };
    if (data) opts.body = JSON.stringify(data);
    const resp = await fetch(url, opts);
    return resp.json();
  }
}
